drivers/adc/adc_ads1119.c

```diff
@@ -0,0 +1,520 @@
+				  struct k_poll_signal *async)
+{
+	int rc;
+	struct ads1119_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	rc = ads1119_adc_start_read(dev, sequence, true);
+	adc_context_release(&data->ctx, rc);
+
+	return rc;
+}
+
+static int ads1119_adc_perform_read(const struct device *dev)
+{
+	int rc;
+	struct ads1119_data *data = dev->data;
+
+	k_sem_take(&data->acq_sem, K_FOREVER);
+
+	rc = ads1119_send_start_read(dev);
+	if (rc) {
+		adc_context_complete(&data->ctx, rc);
+		return rc;
+	}
+
+	rc = ads1119_wait_data_ready(dev);
+	if (rc != 0) {
+		adc_context_complete(&data->ctx, rc);
+		return rc;
+	}
+
+	rc = ads1119_read_sample(dev, data->buffer);
+	if (rc != 0) {
+		adc_context_complete(&data->ctx, rc);
+		return rc;
+	}
+	data->buffer++;
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	return rc;
+}
+
+#if CONFIG_ADC_ASYNC
+static int ads1119_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	return ads1119_adc_read_async(dev, sequence, NULL);
+}
+#else
+static int ads1119_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	int rc;
+	struct ads1119_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	rc = ads1119_adc_start_read(dev, sequence, false);
+
+
+	while (rc == 0 && k_sem_count_get(&data->ctx.sync) == 0) {
+		rc = ads1119_adc_perform_read(dev);
+	}
+	k_sem_take(&data->ctx.sync, K_FOREVER);
+	adc_context_release(&data->ctx, rc);
+	return rc;
+}
+#endif
+
+#if CONFIG_ADC_ASYNC
+static void ads1119_acquisition_thread(struct device *dev)
+{
+	while (true) {
+		ads1119_adc_perform_read(dev);
+	}
+}
+#endif
+
+static int ads1119_init(const struct device *dev)
+{
+	int rc;
+	uint8_t status;
+	const struct ads1119_config *config = dev->config;
+	struct ads1119_data *data = dev->data;
+
+	k_sem_init(&data->acq_sem, 0, 1);
+
+	if (!device_is_ready(config->i2c_dev)) {
+		return -ENODEV;
+	}
+
+	rc = ads1119_read_reg(dev, ADS1119_REG_STATUS, &status);
+	if (rc) {
+		LOG_ERR("Could not get %s ID", dev->name);
+		return rc;
+	}
+
+#if CONFIG_ADC_ASYNC
+	const k_tid_t tid =
+		k_thread_create(&data->thread, data->stack,
+				K_KERNEL_STACK_SIZEOF(data->stack),
+				(k_thread_entry_t)ads1119_acquisition_thread,
+				(void *)dev, NULL, NULL,
+				CONFIG_ADC_ADS1119_ASYNC_THREAD_INIT_PRIO,
+				0, K_NO_WAIT);
+	k_thread_name_set(tid, "adc_ads1119");
+#endif
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return rc;
+}
+
+static const struct adc_driver_api api = {
+	.channel_setup = ads1119_channel_setup,
+	.read = ads1119_read,
+	.ref_internal = 2048,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = ads1119_adc_read_async,
+#endif
+};
+
+#define ADC_ADS1119_INST_DEFINE(n)				     \
+	static const struct ads1119_config config_##n = {	     \
+		.i2c_dev = DEVICE_DT_GET(DT_INST_BUS(n)),	     \
+		.i2c_dt_device_addr = DT_INST_REG_ADDR(n),	     \
+	};							     \
+								     \
+	static struct ads1119_data data_##n = {			     \
```

Comment:
Found out that initialization use 1024B of ROM probably initializing the context in `ads1119_init` will use less ressources

Reply:
It's because the stack is allocated inside the ads1119_data structure. moving it outside should reduce the amount of ROM needed to initialize the structure